pkg/session/session.go

```diff
@@ -208,15 +220,53 @@ func newClient(ctx context.Context, logger logr.Logger, sessionKey string, url *
 	return c, nil
 }
 
-func clearCache(sessionKey string) {
-	sessionMU.Lock()
-	defer sessionMU.Unlock()
-	delete(sessionCache, sessionKey)
+func clearCache(logger logr.Logger, sessionKey string) {
+	if cachedSession, ok := sessionCache.Load(sessionKey); ok {
+		s := cachedSession.(*Session)
+		session, err := s.TagManager.Session(context.Background())
+		if err != nil {
+			logger.Error(err, "unable to get tag manager session")
+		}
+		if session != nil {
+			logger.Info("found active tag manager session, logging out")
+			err := s.TagManager.Logout(context.Background())
+			if err != nil {
+				logger.Error(err, "unable to logout tag manager session")
+			}
+		}
+
+		vimSessionActive, err := s.SessionManager.SessionIsActive(context.Background())
+		if err != nil {
+			logger.Error(err, "unable to get vim client session")
+		} else if vimSessionActive {
+			logger.Info("found active vim session, logging out")
+			err := s.SessionManager.Logout(context.Background())
+			if err != nil {
+				logger.Error(err, "unable to logout vim session")
+			}
+		}
+	}
+	sessionCache.Delete(sessionKey)
 }
 
 // newManager creates a Manager that encompasses the REST Client for the VSphere tagging API.
-func newManager(ctx context.Context, client *vim25.Client, user *url.Userinfo) (*tags.Manager, error) {
+func newManager(ctx context.Context, logger logr.Logger, sessionKey string, client *vim25.Client, user *url.Userinfo, feature Feature) (*tags.Manager, error) {
 	rc := rest.NewClient(client)
+	if feature.EnableKeepAlive {
+		rc.Transport = keepalive.NewHandlerREST(rc, feature.KeepAliveDuration, func() error {
+			s, err := rc.Session(ctx)
+			if err != nil {
+				return err
+			}
+			if s != nil {
+				return nil
+			}
```

Comment:
The documentation of `.Session()` method states `nil` if the session is not authenticated. For us, in case of `err != nil` or `s == nil`, the session is unusable and should be cleared from cache. 
Apart from that, we also want the goroutine to be killed in case of an unusable session, read [doc](https://github.com/vmware/govmomi/blob/6209be5b5c0bd5d81078fdc82eb4001f202f90e7/session/keepalive/handler.go#L63), which would require us to send back an error.

Reply:
Let me test this with a cluster creation, verify and update back

Reply:
@srm09 this implementation is identical to https://github.com/vmware/govmomi/blob/6209be5b5c0bd5d81078fdc82eb4001f202f90e7/session/keepalive/handler.go#L85 do we need to send error  after clearcache() https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/54c7e2f93334d0638144150b077f9783e4cddf61/pkg/session/session.go#L265-L267 here